README.md
# Cuid2

[![Pub Version](https://img.shields.io/pub/v/cuid2?color=cyan)](https://pub.dev/packages/cuid2)
[![Pub Publisher](https://img.shields.io/pub/publisher/cuid2)](https://pub.dev/publishers/obsidia.io/packages)
[![License](https://img.shields.io/badge/license-MIT-blueviolet)](https://github.com/obsidiaHQ/cuid2/blob/master/LICENSE)

Secure, collision-resistant ids optimized for horizontal scaling and performance. Next generation UUIDs.

Need unique ids in your app? Forget UUIDs and GUIDs which often collide in large apps. Use Cuid2, instead.

**Cuid2 is:**

* **Secure:** It's not feasible to guess the next id, existing valid ids, or learn anything about the referenced data from the id. Cuid2 uses multiple, independent entropy sources and hashes them with a security-audited, NIST-standard cryptographically secure hashing algorithm (Sha3).
* **Collision resistant:** It's extremely unlikely to generate the same id twice (by default, you'd need to generate roughly 4,000,000,000,000,000,000 ids ([`sqrt(36^(24-1) * 26) = 4.0268498e+18`](https://en.wikipedia.org/wiki/Birthday_problem#Square_approximation)) to reach 50% chance of collision.
* **Horizontally scalable:** Generate ids on multiple machines without coordination.
* **Offline-compatible:** Generate ids without a network connection.
* **URL and name-friendly:** No special characters.
* **Fast and convenient:** No async operations. Won't introduce user-noticeable delays. Less than 5k, gzipped.
* **But not *too fast*:** If you can hash too quickly you can launch parallel attacks to find duplicates or break entropy-hiding. For unique ids, the fastest runner loses the security race.


**Cuid2 is not good for:**

* Sequential ids (see the [note on K-sortable ids](https://github.com/paralleldrive/cuid2#note-on-k-sortablesequentialmonotonically-increasing-ids), below)
* High performance tight loops, such as render loops (if you don't need cross-host unique ids or security, consider a simple counter for this use-case, or try [Ulid](https://github.com/ulid/javascript) or [NanoId](https://github.com/ai/nanoid)).


**[Learn more](https://github.com/paralleldrive/cuid2)**

## Install
```
dart pub add cuid2
```


## Usage

```dart
import 'package:cuid2/cuid2.dart';

void main() {
  final id = cuid();  // default options
  final id30 = cuidSecure(30);  // set length to 30, use Random.secure()
  final cc = cuidConfig(length: 30);  // custom config - see example

  print(cc.gen())
  print(id); // eh82waoo5fi41lgncwv5oxxb
  print(id30); // oxjkyfqo3aqk3jigelnuyp3ef299qx
}
```

## Testing

a histogram analysis is done on every batch of tests to ensure a fair and random distribution across the whole entropy range. Any bias found during the analysis would increase the chances of ID collision and cause the tests to fail automatically.

![Histogram](https://i.imgur.com/07X7moo.png)

CHANGELOG.md
## 1.0.0

- Initial version.

## 2.0.0

- Add `cuidConfig()` for more customizability.
- Example:

```dart
Function myCounter(int start) {
    return () => start += 5;
}

final cc = cuidConfig(counter: myCounter(0));
final id = cc.gen();
```

- **[BREAKING]** remove `entropyLength` param from `cuid()` and `cuidSecure()`

## 3.0.0

- refactor code and bring it up to cuid2,js v2.2.0
- add more tests
- **[POSSIBLE BREAKING CHANGE]** `counter` parameter function is now of type `int Function()?` instead of `Function?`

## 3.1.0

- fix fingerprinting on web
- increase min SDK version from 2.12.0 to 2.14.0


test/utils.dart
import 'package:cuid2/cuid2.dart';

void info(txt) {
  print('# - $txt');
}

BigInt idToBigInt(String id, {radix = 36}) {
  return id.split('').fold(
      BigInt.zero,
      (r, v) =>
          r * BigInt.from(radix) + BigInt.from(int.parse(v, radix: radix)));
}

Iterable<int> createHistogram(Iterable<BigInt> data, {int bucketCount = 20}) {
  BigInt max = data.reduce((curr, next) => curr > next ? curr : next);
  BigInt interval = BigInt.from((max / BigInt.from(bucketCount)));
  List<Map<String, dynamic>> bins = [];

  // Setup Bins
  for (BigInt i = BigInt.zero; i < BigInt.from(bucketCount); i += BigInt.one) {
    bins.add(
        {'min': i * interval, 'max': interval + (i * interval), 'count': 0});
  }

  for (var number in data) {
    for (var j = 0; j < bins.length; j++) {
      var bin = bins[j];
      if (number > bin['min'] && number <= bin['max']) {
        bin['count']++;
        break; // An item can only be in one bin.
      }
    }
  }

  return bins.map((e) => e['count'] as int);
}

Future<Map<String, dynamic>> createIdPool({max = 100000}) async {
  final set = <String>{};

  for (var i = 0; i < max; i++) {
    set.add(cuid());
    if (i % 10000 == 0) print('${(i / max * 100).floor()}%');
    if (set.length < i) {
      info('Collision at: $i');
      break;
    }
  }
  info('No collisions detected');

  final ids = [...set];
  final numbers = ids.map((x) => idToBigInt(x.substring(1)));
  final histogram = createHistogram(numbers);
  return {'ids': ids, 'numbers': numbers, 'histogram': histogram};
}


test/cuid2_test.dart
import 'dart:typed_data';

import 'package:cuid2/cuid2.dart';
import 'package:test/test.dart';

void main() {
  final id24 = cuid();
  final id16 = cuid(16);
  final id30 = cuid(30);
  final id30s = cuidSecure(30);

  test('must not be empty', () {
    expect(id24, isNotEmpty);
    expect(id30, isNotEmpty);
    expect(id30s, isNotEmpty);
  });

  test('must have required length', () {
    expect(id16.length, equals(16));
    expect(id24.length, equals(24));
    expect(id30.length, equals(30));
    expect(id30s.length, equals(30));
  });

  test('custom fingerprint is set', () {
    String myfingerprint() {
      return "klsdflsdflvcxlkaweporweqwqs";
    }

    final cuid = cuidConfig(fingerprint: myfingerprint);
    expect(cuid.fingerprint!(), equals(myfingerprint()));

    expect(cuidConfig().fingerprint!(), isNotEmpty);
  });

  test('custom counter increments count', () {
    int Function() myCounter(int start) => () => start += 5;

    final cuid = cuidConfig(counter: myCounter(0));
    expect([cuid.counter!(), cuid.counter!(), cuid.counter!()],
        equals([5, 10, 15]));
  });

  test('must be a valid cuid', () {
    expect(isCuid(cuid()), isTrue);
    expect(isCuid('$cuid()$cuid()'), isFalse);
  });

  test('bufToBigInt should return 0', () {
    expect(cuidConfig().bufToBigInt(Uint8List(2)), equals(BigInt.zero));
  });

  test('bufToBigInt should return 2^32 - 1', () {
    expect(
        cuidConfig().bufToBigInt(Uint8List.fromList([0xff, 0xff, 0xff, 0xff])),
        equals(BigInt.from(4294967295)));
  });
}


test/distribution_test.dart
import 'dart:convert';

import 'package:test/test.dart';
import './utils.dart';

void main() async {
  int n = 300000;
  info('Testing $n unique IDs...');
  final pool = await createIdPool(max: n);
  List<String> ids = pool['ids'];
  final sampleIds = ids.sublist(0, 10);
  final set = Set<String>.from(ids);
  final histogram = pool['histogram'];
  info('sample ids: $sampleIds');
  info('histogram: ${histogram.join(",")}');
  final expectedBinSize = (n / histogram.length).ceil();
  final tolerance = 0.05;
  final minBinSize = (expectedBinSize * (1 - tolerance)).round();
  final maxBinSize = (expectedBinSize * (1 + tolerance)).round();
  info('expectedBinSize: $expectedBinSize');
  info('minBinSize: $minBinSize');
  info('maxBinSize: $maxBinSize');

  test('produce a histogram within distribution tolerance', () {
    expect(histogram.every((x) => x > minBinSize && x < maxBinSize), isTrue);
  });

  test('contain only valid characters', () {
    expect(ids.every((id) => RegExp(r'^[a-z0-9]+$').hasMatch(id)), isTrue);
  });

  test('generate no collisions', () {
    expect(set.length, equals(n));
  });

  final idLength = 23;
  final totalLetters = idLength * n;
  final base = 36;
  final toleranceChars = 0.1;
  final expectedBinSizeChars = (totalLetters / base).ceil();
  final minBinSizeChars = (expectedBinSizeChars * (1 - toleranceChars)).round();
  final maxBinSizeChars = (expectedBinSizeChars * (1 + toleranceChars)).round();
  // Drop the first character because it will always be a letter, making
  // the letter frequency skewed.
  final testIds = ids.map((id) => id.substring(2)).toList();
  final Map<String, int> charFrequencies = {};
  testIds.forEach((id) {
    id.split('').forEach((char) {
      charFrequencies[char] = (charFrequencies[char] ?? 0) + 1;
    });
  });

  info("Testing character frequency...");
  info('expectedBinSize: $expectedBinSizeChars');
  info('minBinSize: $minBinSizeChars');
  info('maxBinSize: $maxBinSizeChars');
  info('charFrequencies: ${jsonEncode(charFrequencies)}');

  test('produce even character frequency', () {
    expect(
        charFrequencies.values
            .every((x) => x > minBinSizeChars && x < maxBinSizeChars),
        isTrue);
  });

  test('represent all character values', () {
    expect(charFrequencies.keys.length, base);
  });
}


lib/cuid2.dart
// Copyright 2023 (c) George Mamar
//
// Licensed under MIT license
// ...

/// `cuid2` is a [Dart](https://dart.dev/) implementation of the [cuid2](https://github.com/paralleldrive/cuid2) library.
///
/// For technical details is recommended to check the original documentation where the inner details are explained.
///
/// Example usage:
///
/// ```dart
///   // cuid generator
///   print(cuid()); // => cl8qj639w000l7n6e6p689uad
///
///   // cuid using secure random generator
///   print(cuidSecure()); // => cl8qjcei8000kfq6e3a1q19os
///
///   // custom id length (defaults to 24)
///   print(cuid(30)); // => oxjkyfqo3aqk3jigelnuyp3ef299qx
///
///   final myCuid = cuidConfig(length: 30);
///   final id = myCuid.gen();
/// ```
library cuid2;

import 'src/cuid2_base.dart' show Cuid;

/// Generates a new id (default length of 24).
String cuid([int length = 24]) => Cuid(length).gen();

/// Generates a new id using cryptographic random generator (default length of 24).
///
/// Falls back to `Random()` if the system can not provide a secure source of random numbers. Check [Random.secure()]
String cuidSecure([int length = 24]) => Cuid(length, true).gen();

/// Creates a generator with the given options.
///
/// `length` _optional_ sets the length of the generated IDs (default is 24).
///
/// `counter` _optional_ custom session counter function. Ex. with increments of 5:
/// ```dart
///   myCounter(int start) => () => count+=5;
///   final config = cuidConfig(counter: myCounter(3));
/// ```
/// `fingerprint` _optional_ custom fingerprint function:
/// ```dart
///   String myfingerprint() => "uniqueFingerprint123";
///   final config = cuidConfig(fingerprint: myfingerprint);
/// ```
///
/// `secure` _optional_ uses [Random.secure()] when set to true (if available), otherwise (default & fallback) it uses [Random()]
Cuid cuidConfig(
        {int length = 24,
        int Function()? counter,
        String Function()? fingerprint,
        bool secure = false}) =>
    Cuid(length, secure, counter, fingerprint);

/// Determines if the given string `id` is a valid CUID based on the specified length constraints.
///
/// The function checks if the input string `id` consists only of alphanumeric lowercase characters and
/// if its length is within the given range. The minimum and maximum length constraints can be
/// customized using the optional named parameters `minLength` and `maxLength`, which default to 2 and 32, respectively.
///
/// Returns `true` if the input string `id` is a valid CUID, otherwise returns `false`.
///
/// Example usage:
///
/// ```dart
/// bool isValid = isCuid('abc123');
/// ```
///
/// `id`: The input string to check for validity.
/// `minLength`: The minimum length constraint for the input string (default: 2).
/// `maxLength`: The maximum length constraint for the input string (default: 32).
bool isCuid(String id, {int minLength = 2, int maxLength = 32}) {
  final length = id.length;
  final regex = RegExp(r'^[0-9a-z]+$');

  if (length >= minLength && length <= maxLength && regex.hasMatch(id)) {
    return true;
  }

  return false;
}


lib/src/cuid2_base.dart
import 'dart:convert';
import 'dart:math';
import 'dart:typed_data';
import './fp_native.dart' if (dart.library.html) './fp_web.dart';

import 'package:pointycastle/digests/sha3.dart';

class Cuid {
  static final _cache = <String, Cuid>{};

  int idLength = 24;
  final int _entropyLength = 32;
  int Function()? counter;
  Random _random = Random();
  String Function()? fingerprint;

  // ~22k hosts before 50% chance of initial counter collision
  // with a remaining counter range of 9.0e+15 in JavaScript.
  final int initialCountMax = 476782367;

  final List<String> _alphabet =
      List.generate(26, (i) => String.fromCharCode(i + 97));

  Cuid._create(
      [this.idLength = 24,
      bool secure = false,
      int Function()? counter,
      String Function()? fingerprint]) {
    if (secure) {
      try {
        _random = Random.secure();
      } on UnsupportedError {
        print("Random.secure() is not supported, but 'secure' flag was true");
      }
    }

    this.fingerprint = fingerprint ?? createFingerprint;
    this.counter = counter ??
        _createCounter((_random.nextDouble() * initialCountMax).floor());
  }

  factory Cuid(
      [int idLength = 24,
      bool secure = false,
      int Function()? counter,
      String Function()? fingerprint]) {
    return _cache.putIfAbsent(
        '$secure$idLength${counter.hashCode}${fingerprint.hashCode}',
        () => Cuid._create(idLength, secure, counter, fingerprint));
  }

  String _createEntropy({int length = 4}) {
    String entropy = "";

    while (entropy.length < length) {
      entropy =
          "$entropy${(_random.nextDouble() * 36).floor().toRadixString(36)}";
    }

    return entropy;
  }

  /// Adapted from https://github.com/juanelas/bigint-conversion
  /// MIT License Copyright (c) 2018 Juan Hernández Serrano
  BigInt bufToBigInt(Uint8List buf) {
    final bits = 8;

    BigInt value = BigInt.zero;
    for (final i in buf) {
      final bi = BigInt.from(i);
      value = (value << bits) + bi;
    }
    return value;
  }

  String _hash([String input = ""]) {
    final d = SHA3Digest(512);
    final sha3 = d.process(utf8.encode(input) as Uint8List);

    // Drop the first character because it will bias the histogram
    // to the left.
    return bufToBigInt(sha3).toRadixString(36).substring(1);
  }

  String _randomLetter() {
    return _alphabet[(_random.nextDouble() * _alphabet.length).floor()];
  }

  int Function() _createCounter(int count) => () => count++;

  /// Generates a new id based on the instance configuration
  String gen() {
    final firstLetter = _randomLetter();
    final time = DateTime.now().millisecondsSinceEpoch.toRadixString(36);
    String count = counter!().toRadixString(36);

    // The salt should be long enough to be globally unique across the full
    // length of the hash. For simplicity, we use the same length as the
    // intended id output.
    final salt = _createEntropy(length: idLength);
    final hashInput =
        "$time$salt$count${_hash(fingerprint!()).substring(0, _entropyLength)}";

    return "$firstLetter${_hash(hashInput).substring(1, idLength)}";
  }
}
